### PR TITLE
Sequencer set contract interact with scheduler

### PR DESCRIPTION
**Pull request by Kukoomomo**
Closes: #55

### CODE REVIEW

File: scheduler/drivers/scheduler/driver.go [hunk 1]
```diff
@@ -21,44 +23,66 @@ type Config struct {
...
+		}
+		addr := crypto.PubkeyToAddress(*ecdsaPubKey)
+		deposit, err := tshContract.GetDeposits(nil, common.BytesToAddress(addr.Bytes()))
+
 		seq := &Sequencer{
-			Address: common2.Address(sequencer),
+			Address: addr,
+			PubKey:  *ecdsaPubKey,
+			// todo : caculate voting power
+			VotingPower: deposit.Amount.Int64(),
```
Comment (Kukoomomo) [inline scheduler/drivers/scheduler/driver.go:75]: TODO: How to calculate voting power

File: scheduler/drivers/scheduler/driver.go [hunk 2]
```diff
@@ -21,44 +23,66 @@ type Config struct {
...
+	}
 	var seqz []*Sequencer
-	for _, sequencer := range sequencers {
+	for _, v := range pubKeys {
+		ecdsaPubKey, err := crypto.UnmarshalPubkey(v)
+		if err != nil {
+			return nil, err
+		}
+		addr := crypto.PubkeyToAddress(*ecdsaPubKey)
+		deposit, err := tshContract.GetDeposits(nil, common.BytesToAddress(addr.Bytes()))
```
Comment (Kukoomomo) [inline scheduler/drivers/scheduler/driver.go:69]: To be determined : Contract provide the method for scheduler query a batch of deposit

File: scheduler/drivers/scheduler/driver.go [hunk 3]
```diff
@@ -74,17 +98,31 @@ func (d *Driver) Metrics() metrics.Metrics {
...
+	}
 	var seqz []*Sequencer
-	for _, sequencer := range updateSequencers {
+	for _, v := range pubKeys {
+		ecdsaPubKey, err := crypto.UnmarshalPubkey(v)
+		if err != nil {
+			return err
+		}
+		addr := crypto.PubkeyToAddress(*ecdsaPubKey)
+		deposit, err := d.tshContract.GetDeposits(nil, common.BytesToAddress(addr.Bytes()))
```
Comment (Kukoomomo) [inline scheduler/drivers/scheduler/driver.go:113]: dtto.

File: scheduler/drivers/scheduler/driver.go [hunk 4]
```diff
@@ -74,17 +98,31 @@ func (d *Driver) Metrics() metrics.Metrics {
...
+
 		seq := &Sequencer{
-			Address: common2.Address(sequencer),
+			Address: addr,
+			PubKey:  *ecdsaPubKey,
+			// todo : caculate voting power
+			VotingPower: deposit.Amount.Int64(),
```
Comment (Kukoomomo) [inline scheduler/drivers/scheduler/driver.go:119]: dtto.

File: scheduler/config.go [hunk 5]
```diff
@@ -66,6 +66,12 @@ type Config struct {
 	// SCCAddress is the SCC contract address.
 	SCCAddress string
 
+	// TGMAddressFlag is the TGM contract address.
```
Comment (chengwenxi) [inline scheduler/config.go:69]:
```suggestion
	// TGMAddressFlag is the TSS group manger contract address.
```

File: scheduler/config.go [hunk 6]
```diff
@@ -66,6 +66,12 @@ type Config struct {
...
+	TGMAddress string
+
+	// TSHAddressFlag is the TSH contract address.
```
Comment (chengwenxi) [inline scheduler/config.go:72]:
```suggestion
	// TSHAddressFlag is the TSS staking & slashing contract address.
```

File: scheduler/drivers/scheduler/driver.go (hunk 1 above)
Comment (chengwenxi) [inline scheduler/drivers/scheduler/driver.go:75]: The amount unit is `int256` and cannot be directly converted to `int64`

File: scheduler/drivers/scheduler/driver.go [hunk 7]
```diff
@@ -21,44 +23,66 @@ type Config struct {
...
 	return &Driver{
-		cfg:               cfg,
-		sequencerContract: sequencerContract,
-		seqz:              NewSequencerSet(seqz),
-		metrics:           metrics.NewBase("batch_submitter", cfg.Name),
+		cfg:         cfg,
+		tgmContract: tgmContract,
+		tshContract: tshContract,
+		seqz:        NewSequencerSet(seqz),
+		metrics:     metrics.NewBase("batch_submitter", cfg.Name),
```
Comment (chengwenxi) [inline scheduler/drivers/scheduler/driver.go:85]:
```suggestion
		metrics:     metrics.NewBase("scheduler", cfg.Name),
```

File: scheduler/drivers/scheduler/driver.go [hunk 8]
```diff
@@ -94,3 +132,25 @@ func (d *Driver) SwitchProducer(ctx context.Context) error {
...
+	var changes []*Sequencer
+	unchangeIndex := []int{}
+	for i, v := range new {
+		for _, seq := range d.seqz.Sequencers {
+			if bytes.Equal(seq.Address.Bytes(), v.Address.Bytes()) && v.VotingPower == seq.VotingPower {
+				unchangeIndex = append(unchangeIndex, i)
+				break
+			}
+		}
+	}
```
Comment (chengwenxi) [inline scheduler/drivers/scheduler/driver.go:147]:
```suggestion
	for i, v := range new {
	    changed := true
		for _, seq := range d.seqz.Sequencers {
			if bytes.Equal(seq.Address.Bytes(), v.Address.Bytes()) && v.VotingPower == seq.VotingPower {
				unchangeIndex = append(unchangeIndex, i)
				changed = false
				break
			}
		}
		if changed {
		    changes = append(changes, new[i])
		}
	}
```

**Review by chengwenxi — APPROVED** [review]

**Review by Voidkai — APPROVED** [review]

**Pull request merged**